### PR TITLE
[QA] Add unit tests for core managers and simulation APIs

### PR DESCRIPTION
**Pull request by mayaparov**
## Summary

Add comprehensive unit tests for core OpenCDA managers and simulation APIs. Tests are fully isolated from heavy external dependencies (CARLA, SUMO, torch, open3d) and can run in CI without any simulation infrastructure.

This PR establishes testing patterns and infrastructure that can be reused for future test development.

## Motivation

Previously, these core modules had no unit test coverage, making refactoring risky and regression detection difficult. The modules handle critical functionality:
- **VehicleManager / RSUManager** — lifecycle management of CAVs and roadside units
- **sim_api.ScenarioManager** — CARLA simulation setup and actor spawning
- **cosim_api.CoScenarioManager** — CARLA-SUMO co-simulation synchronization

### New test files

| File | Tests | Description |
|------|-------|-------------|
| `test/test_vehicle_manager.py` | 25 | VehicleManager ID generation, component delegation, lifecycle methods, error propagation |
| `test/test_rsu_manager.py` | 17 | RSUManager ID generation, DataDumper integration, lifecycle methods, error propagation |
| `test/test_sim_api.py` | 31 | ScenarioManager blueprint filtering, weather setup, actor spawning, factory methods, traffic spawning |
| `test/test_cosim_api.py` | 25 | CoScenarioManager actor synchronization, traffic lights, cleanup, `tick()` scenarios, `__init__` smoke |

#### `test/conftest.py`
- `sys.modules` stubs for `carla` and OpenCDA internal modules (injected before import)
- Fixtures with automatic application:
  - `reset_vehicle_manager_state` and `reset_rsu_manager_state` to reset static class variables between tests
- Reusable fixtures: `mock_cav_world`, `minimal_vehicle_config`, `minimal_rsu_config`
- Stub for `omegaconf` if not installed
- Additional lightweight stubs for missing heavy deps for tests under `test/` (e.g. `torch`) to avoid collection/import errors (e.g. `test/test_ml_manager.py` importing production code that does `import torch`)

#### `test/mocked_carla.py` (enhanced)
- **`_FloatAttrMixin`** — base class providing `__eq__` with `math.isclose` and configurable tolerances (`_rel_tol`, `_abs_tol`)
- **`__repr__`** methods for all geometry classes (easier debugging)
- **`__eq__`** methods using `math.isclose` (prevents flaky float comparisons)
- **Deterministic `Vehicle`** — uses `_DEFAULT_CORNERS` constant instead of `np.random.rand()`
- **Backward-compatible signatures**:
  - `Transform(location, rotation)` and `Transform(x, y, z, pitch=0, yaw=0, roll=0)`
  - `BoundingBox(corners)` and `BoundingBox(location, extent)`
- **Strict validation** — unexpected kwargs raise `TypeError`
- **Proper operator protocol** — `Location.__add__` returns `NotImplemented` for type mismatches

### CI configuration
- Added `pytest-mock` to `.github/workflows/requirements/requirements-ci.txt`

### VehicleManager (`test_vehicle_manager.py`)

| Category | Tests |
|----------|-------|
| ID from config | Valid ID, duplicate detection, negative ID handling |
| ID autogeneration | Missing ID, invalid type, duplicate with autogen enabled |
| Prefix handling | `cav`, `platoon`, `unknown`, invalid prefix fallback |
| Delegation | `set_destination` → agent, `update_info` call chain |
| Lifecycle | `run_step` returns control, `destroy` calls all components |
| Edge cases | Platoon application uses `PlatooningBehaviorAgent`, `update_info_v2x` no-op verification |
| Error propagation | localizer/perception/agent/controller failures propagate and stop downstream calls |

### RSUManager (`test_rsu_manager.py`)

| Category | Tests |
|----------|-------|
| ID management | Valid/duplicate/negative/missing ID, autogeneration |
| DataDumper | Created when `data_dumping=True`, `None` otherwise |
| Update cycle | `localize()` and `detect()` called in order |
| Lifecycle | `run_step` with/without dumper, `destroy` cleanup |
| Edge cases | `update_info_v2x` no-op verification |
| Error propagation | localizer/perception failures propagate and stop downstream calls |

### ScenarioManager (`test_sim_api.py`)

| Category | Tests |
|----------|-------|
| Blueprint filtering | `car_blueprint_filter` for 0.9.14/0.9.15, unsupported version exits |
| Multi-class filtering | `multi_class_vehicle_blueprint_filter` by label |
| Weather | `set_weather` creates correct `WeatherParameters` |
| Actor spawning (`spawn_custom_actor`) | model from params, fallback model, color attribute, `color=None`, invalid color type, color not supported (`IndexError` handling + warning) |
| Factory methods | `create_vehicle_manager` empty/single CAV |
| Platoons | `create_platoon_manager` (lead/member wiring), `spawn_special` via `map_helper`, multiple platoons mapping, spawn_position→Transform mapping |
| Traffic | `spawn_vehicles_by_list` and `spawn_vehicle_by_range` (random/non-random, multi-class, fallback to `blueprint_library.filter("vehicle.*")`, lane-change percentages, try_spawn_actor==None skip) |
| Lifecycle | `tick` calls `world.tick()` (future-proofed with `reset_mock()`), `close` restores settings |
| Validation | `sync_mode=False` triggers `sys.exit` |

### CoScenarioManager (`test_cosim_api.py`)

| Category | Tests |
|----------|-------|
| Traffic lights | `traffic_light_ids`, `get_traffic_light_state` exists/missing, traffic-light sync in `tick()` (including `state=None`) |
| Actor spawning | `spawn_actor` success/failure (`INVALID_ACTOR_ID`) |
| Synchronization | `synchronize_vehicle` exists/missing |
| Cleanup | `destroy_actor` exists/missing, `close` full cleanup |
| `tick()` SUMO→CARLA | spawn/destroy/update mappings; blueprint missing → unsubscribe; spawn failure does not create mapping |
| `tick()` CARLA→SUMO | spawn/destroy mappings; unknown node id fallback (`unknown-<id>`); ignore non `vehicle.*`/`static.*`; do not spawn when `get_sumo_vtype()` returns `None` |
| Init smoke (`__init__`) | initialization with fake `*.sumocfg` + mocked `SumoSimulation`; missing `*.sumocfg` asserts |

### 1. Complete isolation from external dependencies
All heavy dependencies are mocked via `sys.modules` injection **before** any OpenCDA module is imported:
```python
# In conftest.py (runs before test collection)
sys.modules["carla"] = carla_stub
```

### 2. Deterministic test data
- No `np.random` without explicit control
- Static class variables (`VehicleManager.used_ids`, etc.) reset via automatically-applied fixtures
- `_DEFAULT_CORNERS` constant replaces random bounding box generation

### 3. Strict mock contracts
- `spec_set` used for mocks where appropriate to catch unexpected method calls
- `reset_mock()` called before assertions to isolate from `__init__` side effects
- `ruff check test/` is clean (used to catch accidental redefinitions / unused variables)

### 4. Float comparison safety
```python
class _FloatAttrMixin:
    _cmp_attrs: tuple = ()
    _rel_tol: float = 1e-9
    _abs_tol: float = 1e-9

    def __eq__(self, other):
        return all(
            math.isclose(getattr(self, a), getattr(other, a), rel_tol=self._rel_tol, abs_tol=self._abs_tol)
            for a in self._cmp_attrs
        )
```

### 5. Backward compatibility
`mocked_carla.py` maintains full backward compatibility with existing tests by supporting multiple call signatures.

### 6. `CoScenarioManager.tick()` tested by scenarios, not one monolithic test
`tick()` is validated via multiple small tests (spawn/destroy/update paths, traffic lights sync, fallback/error paths). This keeps tests unit-level and stable while still covering critical branches.

## How to run
```bash
# All new tests
pytest test/test_vehicle_manager.py test/test_rsu_manager.py test/test_sim_api.py test/test_cosim_api.py -v

# Quick check
pytest test/test_vehicle_manager.py test/test_rsu_manager.py test/test_sim_api.py test/test_cosim_api.py -q

# With coverage (if pytest-cov installed)
pytest test/test_vehicle_manager.py test/test_rsu_manager.py test/test_sim_api.py test/test_cosim_api.py --cov=opencda
```

### CODE REVIEW

File: test/test_cosim_api.py [hunk 1]
```diff
@@ -0,0 +1,693 @@
...
+
+import pytest
+
+from test import mocked_carla as carla
+
+
+def _make_cosim_manager_without_init():
+    from opencda.scenario_testing.utils.cosim_api import CoScenarioManager
+
+    mgr = CoScenarioManager.__new__(CoScenarioManager)
```
Comment (Haralishev77) [inline test/test_cosim_api.py]: Let's use `__init__` instead of `__new__`
